numba/core/event.py

```diff
@@ -0,0 +1,420 @@
+"""
+The ``numba.core.event`` module provides a simple event system for applications
+to register callbacks to listen to specific compiler events.
+
+The following events are built in:
+
+- ``"numba:compile"`` is broadcast when a dispatcher is compiling. Events of
+  this kind have ``data`` defined to be a ``dict`` with the following
+  key-values:
+
+  - ``"dispatcher"``: the dispatcher object that is compiling.
+  - ``"args"``: the argument types.
+  - ``"return_type"``: the return type.
+
+- ``"numba:compiler_lock"`` is broadcast when the internal compiler-lock is
+  acquired. This is mostly used internally to measure time spent with the lock
+  acquired.
+
+Applications can register callbacks that are listening for specific events using
+``register(kind: str, listener: Listener)``, where ``listener`` is an instance
+of ``Listener`` that defines custom actions on occurrence of the specific event.
+"""
+
+import abc
+import enum
+import time
+from timeit import default_timer as timer
+from contextlib import contextmanager, ExitStack
+from collections import defaultdict
+
+
+class EventStatus(enum.Enum):
+    """Status of an event.
+    """
+    START = enum.auto()
+    END = enum.auto()
+
+
+# Builtin event kinds.
+_builtin_kinds = frozenset([
+    "numba:compiler_lock",
+    "numba:compile",
+])
+
+
+def _guard_kind(kind):
+    """Guard to ensure that an event kind is valid.
+
+    All event kinds with a "numba:" prefix must be defined in the pre-defined
+    ``numba.core.event._builtin_kinds``.
+    Custom event kinds are allowed by not using the above prefix.
+
+    Parameters
+    ----------
+    kind: str
+
+    Return
+    ------
+    res: str
+    """
+    if kind.startswith("numba:") and kind not in _builtin_kinds:
+        msg = (f"{kind} is not a valid event kind, "
+               "it starts with the reserved prefix 'numba:'")
+        raise ValueError(msg)
+    return kind
+
+
+class Event:
+    """An event.
+
+    Parameters
+    ----------
+    kind : str
+    status : EventStatus
+    data : any; optional
+        Additional data for the event.
+    exc_details : 3-tuple; optional
```

Comment:
```suggestion
    kind: str
    status: EventStatus
    data: any; optional
        Additional data for the event.
    exc_details: 3-tuple; optional
```

Reply:
numpydoc rendering issue should be fixed.